File: src/hya/eval.py
"""Stack-based evaluator for Hya.

Evaluation model (Forth-like + Lisp):
1. Expressions are evaluated left-to-right, top-down
2. The evaluator maintains a data stack (list)
3. Literals push themselves onto the stack
4. Builtins pop arguments (according to their arity), compute, push results
5. User-defined functions (Function) create a new lexical scope, bind args,
   then evaluate their body
6. Special forms (if, define, defn, fn, quote, do, let, set!) have custom
   evaluation rules
"""

from __future__ import annotations
from typing import Any, Optional

from .types import Symbol, nil, Builtin, Function, is_truthy, hya_repr
from .env import Environment
from .parse import ArityTable, Parser, parse_source
from .builtins import get_builtins


# ---------------------------------------------------------------------------
# Evaluator
# ---------------------------------------------------------------------------

class Evaluator:
    """Stack-based Hya evaluator.

    The evaluator walks an AST (nested Python lists produced by the parser)
    and evaluates it left-to-right.

    Attributes:
        stack: The data stack (Forth-like).
        env: The current environment.
        global_env: The root environment.
        arity_table: Table of known arities (for the parser).
        debug: If True, print trace info.
    """

    def __init__(self, debug: bool = False) -> None:
        self.stack: list[Any] = []
        self.global_env = Environment(name="global")
        self.env = self.global_env
        self.arity_table = ArityTable()
        self.debug = debug
        self._parser: Optional[Parser] = None

        # Load builtins
        self._load_builtins()

    @property
    def parser(self) -> Parser:
        if self._parser is None:
            self._parser = Parser(self.arity_table)
        return self._parser

    def _load_builtins(self) -> None:
        """Register all builtins in the environment and arity table."""
        for name, builtin in get_builtins().items():
            self.global_env.define(name, builtin)
            if builtin.arity >= 0:
                self.arity_table.register(name, builtin.arity)
            # arity -1 (variadic) means the symbol is registered but with
            # 'variadic' marker — user must use parens
        # Special form arities — ALL fixed, no parens needed
        self.arity_table.register("define", 2)   # define name value
        self.arity_table.register("quote", 1)    # quote expr
        self.arity_table.register("do", -1)      # do -> variadic (use parens)
        self.arity_table.register("set!", 2)     # set! name value
        self.arity_table.register("let", 2)      # let bindings body
        self.arity_table.register("if", 3)       # if cond then else
        self.arity_table.register("while", 2)    # while cond body
        self.arity_table.register("for", 3)      # for var list body
        self.arity_table.register("cond", 1)     # cond clauses-list
        self.arity_table.register("defn", 3)     # defn name (params) body
        self.arity_table.register("fn", 2)       # fn (params) body
        self.arity_table.register("import", 1)   # import module-name
        self.arity_table.register("import!", 2)  # import! module alias
        self.arity_table.register(".", 2)        # . obj attr
        self.arity_table.register("python", 1)   # python "code"

    def eval(self, expr: Any) -> Any:
        """Evaluate a single expression and return the result.

        The result is also pushed onto the data stack.
        """
        result = self._eval_expr(expr)
        # Push result onto the data stack
        if result is not None:
            self.stack.append(result)
        return result

    def _eval_expr(self, expr: Any) -> Any:
        """Internal recursive evaluation of a single expression."""
        if self.debug:
            print(f"  EVAL: {hya_repr(expr)}  stack=[{','.join(hya_repr(e) for e in self.stack[-3:])}]")

        # Literals evaluate to themselves
        if isinstance(expr, (int, float, str)):
            return expr

        if expr is nil:
            return nil

        if expr is True:
            return True

        if expr is False:
            return False

        # Symbol: look up in environment (handle special constants)
        if isinstance(expr, Symbol):
            name = expr.name
            # Built-in constants
            if name == "nil":
                return nil
            if name == "true":
                return True
            if name == "false":
                return False
            val = self.env.get(name)
            if val is None:
                raise NameError(f"Undefined symbol: {name}")
            return val

        # List: S-expression application
        if isinstance(expr, list):
            if not expr:
                return nil

            head = expr[0]

            # ---- Special forms ----

            # QUOTE: return the argument unevaluated
            if isinstance(head, Symbol) and head.name == "quote":
                if len(expr) < 2:
                    raise SyntaxError("quote expects 1 argument")
                return expr[1]

            # DEFINE: bind a name to a value
            if isinstance(head, Symbol) and head.name == "define":
                if len(expr) < 3:
                    raise SyntaxError("define expects (define name value)")
                name_expr = expr[1]
                value_expr = expr[2]
                if isinstance(name_expr, Symbol):
                    name = name_expr.name
                    value = self._eval_expr(value_expr)
                    self.env.define(name, value)
                    # Register arity for function values
                    if isinstance(value, Function):
                        self.arity_table.register(name, value.arity)
                    elif callable(value):
                        # Try to detect arity of Python callables
                        try:
                            import inspect
                            sig = inspect.signature(value)
                            required = sum(
                                1 for p in sig.parameters.values()
                                if p.default is inspect.Parameter.empty
                                and p.kind in (
                                    inspect.Parameter.POSITIONAL_ONLY,
                                    inspect.Parameter.POSITIONAL_OR_KEYWORD))
                            if required >= 0:
                                self.arity_table.register(name, required)
                        except (ValueError, TypeError):
                            pass  # Can't determine arity, skip
                    return value
                raise SyntaxError(
                    f"define expects a symbol name, got {name_expr}")

            # DEFN: define a function with known arity
            if isinstance(head, Symbol) and head.name == "defn":
                if len(expr) < 4:
                    raise SyntaxError(
                        "defn expects (defn name (params) body...)")
                name_sym = expr[1]
                params = expr[2]
                body = expr[3:]
                if not isinstance(name_sym, Symbol):
                    raise SyntaxError(
                        f"defn expects a symbol name, got {name_sym}")
                if not isinstance(params, list):
                    raise SyntaxError(
                        "defn expects a parameter list")
                # Create function
                fn = Function(params, body, self.env, name_sym.name)
                self.env.define(name_sym.name, fn)
                # Arity was registered by parser; ensure it matches
                self.arity_table.register(name_sym.name, len(params))
                return fn

            # IF: conditional with arity 3
            if isinstance(head, Symbol) and head.name == "if":
                if len(expr) < 4:
                    raise SyntaxError("if expects (if cond then else)")
                cond = self._eval_expr(expr[1])
                if is_truthy(cond):
                    return self._eval_expr(expr[2])
                else:
                    return self._eval_expr(expr[3])

            # DO: evaluate multiple exprs, return last
            if isinstance(head, Symbol) and head.name == "do":
                result = nil
                for subexpr in expr[1:]:
                    result = self._eval_expr(subexpr)
                return result

            # FN: create anonymous function
            if isinstance(head, Symbol) and head.name == "fn":
                if len(expr) < 3:
                    raise SyntaxError("fn expects (fn (params) body...)")
                params = expr[1]
                body = expr[2:]
                if not isinstance(params, list):
                    raise SyntaxError(
                        "fn expects a parameter list")
                return Function(params, body, self.env)

            # WHILE: loop while condition is truthy (arity 2)
            if isinstance(head, Symbol) and head.name == "while":
                if len(expr) < 3:
                    raise SyntaxError(
                        "while expects (while cond body)")
                cond_expr = expr[1]
                body_exprs = expr[2:]
                result = nil
                while is_truthy(self._eval_expr(cond_expr)):
                    for subexpr in body_exprs:
                        result = self._eval_expr(subexpr)
                return result

            # FOR: iterate over a list (arity 3: var list body)
            if isinstance(head, Symbol) and head.name == "for":
                if len(expr) < 4:
                    raise SyntaxError(
                        "for expects (for var list body)")
                var_expr = expr[1]
                list_expr = self._eval_expr(expr[2])
                body_expr = expr[3:]
                if not isinstance(var_expr, Symbol):
                    raise SyntaxError(
                        f"for expects a symbol as variable, got {var_expr}")
                if not isinstance(list_expr, list):
                    raise TypeError(
                        f"for expects a list, got {type(list_expr)}")
                result = nil
                for item in list_expr:
                    self.env.define(var_expr.name, item)
                    for subexpr in body_expr:
                        result = self._eval_expr(subexpr)
                return result

            # COND: multi-branch conditional (arity 1: clause-list)
            # Clauses are flat pairs: (test1 result1 test2 result2 ...)
            if isinstance(head, Symbol) and head.name == "cond":
                if len(expr) < 2:
                    raise SyntaxError(
                        "cond expects (cond clause...)")
                clauses = expr[1]
                if isinstance(clauses, list):
                    i = 0
                    while i < len(clauses) - 1:
                        test = self._eval_expr(clauses[i])
                        result = clauses[i + 1]
                        if is_truthy(test):
                            return self._eval_expr(result)
                        i += 2
                    # Trailing single clause (else-like)
                    if i < len(clauses):
                        if is_truthy(self._eval_expr(clauses[i])):
                            return self._eval_expr(clauses[i])
                return nil

            # SET!: mutate a binding
            if isinstance(head, Symbol) and head.name == "set!":
                if len(expr) < 3:
                    raise SyntaxError("set! expects (set! name value)")
                name_expr = expr[1]
                value = self._eval_expr(expr[2])
                if isinstance(name_expr, Symbol):
                    self.env.set(name_expr.name, value)
                    return value
                raise SyntaxError(
                    f"set! expects a symbol name, got {name_expr}")

            # LET: local bindings
            if isinstance(head, Symbol) and head.name == "let":
                if len(expr) < 3:
                    raise SyntaxError(
                        "let expects (let ((name val)...) body...)")
                bindings = expr[1]
                body = expr[2:]
                # Create a new scope
                let_env = self.env.extend("let")
                old_env = self.env
                self.env = let_env
                try:
                    if isinstance(bindings, list):
                        for binding in bindings:
                            if (isinstance(binding, list)
                                    and len(binding) >= 2):
                                bname = binding[0]
                                bval = self._eval_expr(binding[1])
                                if isinstance(bname, Symbol):
                                    self.env.define(bname.name, bval)
                    result = nil
                    for subexpr in body:
                        result = self._eval_expr(subexpr)
                    return result
                finally:
                    self.env = old_env

            # ---- Python Interop ----

            # IMPORT: import a Python module (arity 1: import os)
            if isinstance(head, Symbol) and head.name in ("import", "import!"):
                if len(expr) < 2:
                    raise SyntaxError("import expects (import module-name)")
                module_name = expr[1]
                if isinstance(module_name, Symbol):
                    import importlib
                    try:
                        mod = importlib.import_module(module_name.name)
                        name = module_name.name
                        # Use import! to bind under a different name
                        if head.name == "import!" and len(expr) >= 3:
                            if isinstance(expr[2], Symbol):
                                name = expr[2].name
                        self.env.define(name, mod)
                        return mod
                    except ImportError as e:
                        raise ImportError(
                            f"Cannot import Python module '{module_name.name}': {e}")
                raise TypeError(
                    f"import expects a symbol, got {module_name}")

            # DOT: chained attribute access
            # (. os path join "a" "b") = os.path.join("a", "b")
            # Top-level: . os path (arity 2) = getattr(os, 'path')
            if isinstance(head, Symbol) and head.name == ".":
                if len(expr) < 3:
                    raise SyntaxError(
                        ". expects (. obj attr [attr...] [args...])")
                obj = self._eval_expr(expr[1])
                # Walk through attr chain, then collect call args
                i = 2
                while i < len(expr):
                    item = expr[i]
                    if isinstance(item, Symbol):
                        # Attribute access
                        attr_name = item.name
                        if hasattr(obj, attr_name):
                            obj = getattr(obj, attr_name)
                            i += 1
                        else:
                            raise AttributeError(
                                f"'{type(obj).__name__}' "
                                f"has no attribute '{attr_name}'")
                    else:
                        # Not a symbol — start of call args
                        break
                # If we have remaining items, call the result
                if i < len(expr):
                    call_args = [self._eval_expr(a)
                                 for a in expr[i:]]
                    if callable(obj):
                        return obj(*call_args)
                    raise TypeError(
                        f"Cannot call non-callable: {obj}")
                return obj

            # PYTHON: evaluate arbitrary Python expression
            if isinstance(head, Symbol) and head.name == "python":
                if len(expr) < 2:
                    raise SyntaxError(
                        "python expects (python \"code\")")
                code = expr[1]
                if isinstance(code, str):
                    import builtins as py_builtins
                    # Evaluate in context of the global environment
                    local_vars = {}
                    for k, v in self.global_env._bindings.items():
                        local_vars[k] = v
                    try:
                        result = eval(code, py_builtins.__dict__, local_vars)
                        return result
                    except Exception as e:
                        raise RuntimeError(
                            f"Python eval error: {e}")
                raise TypeError(
                    f"python expects a string, got {type(code)}")

            # ---- Generic function application ----
            if isinstance(head, Symbol) and head.name == "defn-rec":
                # Same as defn but the function body can recurse
                if len(expr) < 4:
                    raise SyntaxError(
                        "defn-rec expects (defn-rec name (params) body...)")
                name_sym = expr[1]
                params = expr[2]
                body = expr[3:]
                fn = Function(params, body, self.env, name_sym.name)
                # Define BEFORE evaluating body (for recursion)
                self.env.define(name_sym.name, fn)
                self.arity_table.register(name_sym.name, len(params))
                return fn

            # Generic function call
            fn_val = self._eval_expr(head)
            args = [self._eval_expr(arg) for arg in expr[1:]]

            if isinstance(fn_val, Builtin):
                return fn_val(*args, evaluator=self)

            if isinstance(fn_val, Function):
                return self._apply_function(fn_val, args)

            if callable(fn_val):
                return fn_val(*args)

            raise TypeError(
                f"Cannot call non-function: {hya_repr(fn_val)}")

        raise TypeError(f"Unknown expression type: {type(expr)}: {expr}")

    def _apply_function(self, fn: Function, args: list) -> Any:
        """Apply a user-defined function with given arguments.

        Creates a new lexical scope, binds parameters to args,
        evaluates the body, and returns the result.
        """
        if len(args) != len(fn.params):
            raise TypeError(
                f"Function expected {len(fn.params)} args, "
                f"got {len(args)}")

        # Create call environment
        call_env = fn.env.extend(f"call({fn.name or 'anon'})")
        old_env = self.env
        self.env = call_env

        try:
            # Bind parameters
            for param, arg in zip(fn.params, args):
                self.env.define(param.name, arg)

            # Evaluate body
            if isinstance(fn.body, list):
                result = nil
                for subexpr in fn.body:
                    result = self._eval_expr(subexpr)
                return result
            else:
                return self._eval_expr(fn.body)
        finally:
            self.env = old_env

    def exec(self, source: str) -> Any:
        """Parse and evaluate Hya source code.

        Expressions are parsed and evaluated ONE AT A TIME so that
        arity registrations from `defn`/`define` take effect for
        subsequent expressions in the same source text.

        Args:
            source: Hya source string.

        Returns:
            Result of the last expression.
        """
        from .tokenize import tokenize
        from .parse import TokenStream
        tokens = tokenize(source)
        stream = TokenStream(tokens)
        result = nil
        while not stream.is_eof:
            expr = self.parser._parse_expr(stream, allow_arity=True)
            if expr is not None:
                result = self.eval(expr)
        return result

    def exec_file(self, path: str) -> Any:
        """Load and execute a .hya file."""
        with open(path, 'r', encoding='utf-8') as f:
            source = f.read()
        return self.exec(source)

File: js/arli/src/eval.js
/**
 * Stack-based evaluator for arli (JavaScript port).
 */

import { ArliSymbol, nil, Builtin, Function, isTruthy, arliRepr, arliEqual } from './types.js';
import { Environment } from './env.js';
import { ArityTable, Parser } from './parse.js';
import { tokenize, TokenStream } from './tokenize.js';
import { getBuiltins } from './builtins.js';

let nodeFs = null;
let nodeRequire = null;
if (typeof process !== 'undefined' && process.versions && process.versions.node) {
  try {
    const { createRequire } = await import('node:module');
    nodeRequire = createRequire(import.meta.url);
    nodeFs = nodeRequire('node:fs');
  } catch (e) {}
}

export class Evaluator {
  constructor(debug = false) {
    this.stack = [];
    this.execStack = [];
    this.envStack = [];
    this.globalEnv = new Environment(null, 'global');
    this.env = this.globalEnv;
    this.arityTable = new ArityTable();
    this.debug = debug;
    this._parser = null;
    this._loadBuiltins();
  }

  get parser() {
    if (this._parser === null) this._parser = new Parser(this.arityTable);
    return this._parser;
  }

  _loadBuiltins() {
    for (const [name, builtin] of Object.entries(getBuiltins())) {
      this.globalEnv.define(name, builtin);
      if (builtin.arity >= 0) this.arityTable.register(name, builtin.arity);
    }
    this.arityTable.register('define', 2);
    this.arityTable.register('quote', 1);
    this.arityTable.register('do', -1);
    this.arityTable.register('set!', 2);
    this.arityTable.register('let', 2);
    this.arityTable.register('if', 3);
    this.arityTable.register('while', 2);
    this.arityTable.register('for', 3);
    this.arityTable.register('cond', 1);
    this.arityTable.register('defn', 3);
    this.arityTable.register('defn-rec', 3);
    this.arityTable.register('fn', 2);
    this.arityTable.register('import', 1);
    this.arityTable.register('import!', 2);
    this.arityTable.register('.', 2);
    this.arityTable.register('host', 1);
    this.arityTable.register('assert', 2);
    this.arityTable.register('doc', 1);
    this.arityTable.register('doc!', 2);
    this.arityTable.register('match', -1);
    this.arityTable.register('import-module', 1);
    this.arityTable.register('defn-fexpr', 3);
    this.arityTable.register('eval', 1);
  }

  evalForm(form) { return this._evalExpr(form); }

  eval(expr) {
    const result = this._evalExpr(expr);
    if (result !== null && result !== undefined) this.stack.push(result);
    return result;
  }

  _evalExpr(expr) {
    if (this.debug) {
      const stackPreview = this.stack.slice(-3).map(arliRepr).join(',');
      console.log(`  EVAL: ${arliRepr(expr)}  stack=[${stackPreview}]`);
    }

    if (typeof expr === 'number' || typeof expr === 'string') return expr;
    if (expr === nil) return nil;
    if (typeof expr === 'boolean') return expr;

    if (expr instanceof ArliSymbol) {
      const name = expr.name;
      if (name.startsWith(':')) return expr;
      if (name === 'nil') return nil;
      if (name === 'true') return true;
      if (name === 'false') return false;
      const val = this.env.get(name);
      if (val === null || val === undefined) throw new ReferenceError(`Undefined symbol: ${name}`);
      return val;
    }

    if (Array.isArray(expr)) {
      if (expr.length === 0) return nil;
      const head = expr[0];

      // Special forms
      if (head instanceof ArliSymbol && head.name === 'quote') {
        if (expr.length < 2) throw new SyntaxError('quote expects 1 argument');
        return expr[1];
      }

      if (head instanceof ArliSymbol && head.name === 'define') {
        if (expr.length < 3) throw new SyntaxError('define expects (define name value)');
        const nameExpr = expr[1], valueExpr = expr[2];
        if (nameExpr instanceof ArliSymbol) {
          const name = nameExpr.name;
          const value = this._evalExpr(valueExpr);
          this.env.define(name, value);
          if (value instanceof Function) this.arityTable.register(name, value.arity);
          return value;
        }
        throw new SyntaxError(`define expects a symbol name, got ${nameExpr}`);
      }

      if (head instanceof ArliSymbol && head.name === 'defn') {
        if (expr.length < 4) throw new SyntaxError('defn expects (defn name (params) body...)');
        const nameSym = expr[1], params = expr[2], body = expr.slice(3);
        if (!(nameSym instanceof ArliSymbol)) throw new SyntaxError(`defn expects a symbol name, got ${nameSym}`);
        if (!Array.isArray(params)) throw new SyntaxError('defn expects a parameter list');
        const fn = new Function(params, body, this.env, nameSym.name);
        this.env.define(nameSym.name, fn);
        this.arityTable.register(nameSym.name, params.length);
        return fn;
      }

      if (head instanceof ArliSymbol && head.name === 'if') {
        if (expr.length < 4) throw new SyntaxError('if expects (if cond then else)');
        const cond = this._evalExpr(expr[1]);
        return this._evalExpr(isTruthy(cond) ? expr[2] : expr[3]);
      }

      if (head instanceof ArliSymbol && head.name === 'do') {
        let result = nil;
        for (const sub of expr.slice(1)) result = this._evalExpr(sub);
        return result;
      }

      if (head instanceof ArliSymbol && head.name === 'fn') {
        if (expr.length < 3) throw new SyntaxError('fn expects (fn (params) body...)');
        const params = expr[1], body = expr.slice(2);
        if (!Array.isArray(params)) throw new SyntaxError('fn expects a parameter list');
        return new Function(params, body, this.env);
      }

      if (head instanceof ArliSymbol && head.name === 'while') {
        if (expr.length < 3) throw new SyntaxError('while expects (while cond body)');
        let result = nil;
        while (isTruthy(this._evalExpr(expr[1]))) { for (const sub of expr.slice(2)) result = this._evalExpr(sub); }
        return result;
      }

      if (head instanceof ArliSymbol && head.name === 'for') {
        if (expr.length < 4) throw new SyntaxError('for expects (for var list body)');
        const varExpr = expr[1];
        const listVal = this._evalExpr(expr[2]);
        const bodyExprs = expr.slice(3);
        if (!(varExpr instanceof ArliSymbol)) throw new SyntaxError(`for expects a symbol as variable, got ${varExpr}`);
        if (!Array.isArray(listVal)) throw new TypeError(`for expects a list, got ${typeof listVal}`);
        let result = nil;
        for (const item of listVal) { this.env.define(varExpr.name, item); for (const sub of bodyExprs) result = this._evalExpr(sub); }
        return result;
      }

      if (head instanceof ArliSymbol && head.name === 'cond') {
        if (expr.length < 2) throw new SyntaxError('cond expects (cond clause...)');
        const clauses = expr[1];
        if (Array.isArray(clauses)) {
          let i = 0;
          while (i < clauses.length - 1) {
            if (isTruthy(this._evalExpr(clauses[i]))) return this._evalExpr(clauses[i + 1]);
            i += 2;
          }
          if (i < clauses.length && isTruthy(this._evalExpr(clauses[i]))) return this._evalExpr(clauses[i]);
        }
        return nil;
      }

      if (head instanceof ArliSymbol && head.name === 'set!') {
        if (expr.length < 3) throw new SyntaxError('set! expects (set! target value)');
        const nameExpr = expr[1];
        const value = this._evalExpr(expr[expr.length - 1]);
        if (nameExpr instanceof ArliSymbol && expr.length === 3) { this.env.set(nameExpr.name, value); return value; }
        if (expr.length >= 4) {
          const obj = this._evalExpr(expr[1]), key = expr[2];
          if (typeof obj === 'object' && obj !== null && !Array.isArray(obj)) {
            if (key instanceof ArliSymbol && key.name.startsWith(':')) obj[key.name] = value;
            else if (typeof key === 'string') obj[key] = value;
            else obj[String(key)] = value;
            return value;
          }
          if (Array.isArray(obj) && typeof key === 'number') { const idx = Math.floor(key); if (idx >= 0 && idx < obj.length) { obj[idx] = value; return value; } }
        }
        throw new SyntaxError(`set! cannot set on target: ${arliRepr(nameExpr)}`);
      }

      if (head instanceof ArliSymbol && head.name === 'let') {
        if (expr.length < 3) throw new SyntaxError('let expects (let ((name val)...) body...)');
        const bindings = expr[1], body = expr.slice(2);
        const letEnv = this.env.extend('let');
        const oldEnv = this.env;
        this.env = letEnv;
        try {
          if (Array.isArray(bindings)) {
            for (const binding of bindings) {
              if (Array.isArray(binding) && binding.length >= 2) {
                const bname = binding[0], bval = this._evalExpr(binding[1]);
                if (bname instanceof ArliSymbol) this.env.define(bname.name, bval);
              }
            }
          }
          let result = nil;
          for (const sub of body) result = this._evalExpr(sub);
          return result;
        } finally { this.env = oldEnv; }
      }

      if (head instanceof ArliSymbol && (head.name === 'import' || head.name === 'import!')) {
        if (expr.length < 2) throw new SyntaxError('import expects (import module-name)');
        let moduleName = '';
        const nameTok = expr[1];
        if (nameTok instanceof ArliSymbol) moduleName = nameTok.name;
        else if (typeof nameTok === 'string') moduleName = nameTok;
        else throw new TypeError(`import expects a symbol or string, got ${nameTok}`);
        // Node: sync require(). Browser: fire dynamic import(), module available when resolved.
        try {
          let name = moduleName;
          if (head.name === 'import!' && expr.length >= 3 && expr[2] instanceof ArliSymbol) name = expr[2].name;
          let mod;
          if (nodeRequire) {
            mod = nodeRequire(moduleName);
            this.env.define(name, mod);
            return mod;
          } else {
            // Browser ESM: dynamic import() — returns a Promise.
            // We return a loading marker; the real module replaces it when resolved.
            const marker = { __module__: moduleName, __loading__: true };
            this.env.define(name, marker);
            import(moduleName)
              .then(m => { this.env.define(name, m); })
              .catch(() => { /* import failed, marker remains */ });
            return marker;
          }
        } catch (e) {
          const marker = { __module__: moduleName, __error__: e.message };
          let name = moduleName;
          if (head.name === 'import!' && expr.length >= 3 && expr[2] instanceof ArliSymbol) name = expr[2].name;
          this.env.define(name, marker);
          return marker;
        }
      }

      if (head instanceof ArliSymbol && head.name === '.') {
        if (expr.length < 3) throw new SyntaxError('. expects (. obj attr [attr...] [args...])');
        let obj = this._evalExpr(expr[1]);
        let i = 2;
        while (i < expr.length) {
          const item = expr[i];
          if (item instanceof ArliSymbol) { obj = obj[item.name]; i++; }
          else break;
        }
        if (i < expr.length) {
          const callArgs = [];
          for (let j = i; j < expr.length; j++) callArgs.push(this._evalExpr(expr[j]));
          if (typeof obj === 'function') return obj(...callArgs);
          throw new TypeError(`Cannot call non-callable: ${obj}`);
        }
        return obj;
      }

      if (head instanceof ArliSymbol && head.name === 'host') {
        if (expr.length < 2) throw new SyntaxError('host expects (host "code")');
        const code = expr[1];
        if (typeof code === 'string') {
          try {
            const fn = new Function(...Object.keys(this.globalEnv._bindings), `return ${code}`);
            return fn(...Object.values(this.globalEnv._bindings));
          } catch (e) { throw new Error(`Host eval error: ${e.message}`); }
        }
        throw new TypeError(`host expects a string, got ${typeof code}`);
      }

      if (head instanceof ArliSymbol && head.name === 'assert') {
        if (expr.length < 2) throw new SyntaxError('assert expects (assert expr message)');
        const val = this._evalExpr(expr[1]);
        if (!isTruthy(val)) {
          let msg = '';
          if (expr.length >= 3) msg = arliRepr(this._evalExpr(expr[2]));
          throw new Error(`Assertion failed: ${msg}`);
        }
        return val;
      }

      if (head instanceof ArliSymbol && head.name === 'doc') {
        if (expr.length >= 2 && expr[1] instanceof ArliSymbol) {
          const docVal = this.env.lookup(`__doc_${expr[1].name}`);
          if (docVal !== null && docVal !== undefined) return docVal;
        }
        return nil;
      }

      if (head instanceof ArliSymbol && head.name === 'doc!') {
        if (expr.length >= 3 && expr[1] instanceof ArliSymbol) {
          const docText = this._evalExpr(expr[2]);
          this.env.define(`__doc_${expr[1].name}`, docText);
          return docText;
        }
        return nil;
      }

      if (head instanceof ArliSymbol && head.name === 'match') {
        if (expr.length < 2) throw new SyntaxError('match expects (match expr clause...)');
        const matchVal = this._evalExpr(expr[1]);
        for (const clause of expr.slice(2)) {
          if (Array.isArray(clause) && clause.length >= 2) {
            if (this._matchPattern(clause[0], matchVal)) return this._evalExpr(clause[1]);
          }
        }
        return nil;
      }

      if (head instanceof ArliSymbol && head.name === 'import-module') {
        if (expr.length < 2) throw new SyntaxError('import-module expects (import-module "path")');
        const path = this._evalExpr(expr[1]);
        if (typeof path === 'string') {
          return this.execFile(path);
        }
        throw new TypeError(`import-module expects a string path`);
      }

      if (head instanceof ArliSymbol && head.name === 'defn-rec') {
        if (expr.length < 4) throw new SyntaxError('defn-rec expects (defn-rec name (params) body...)');
        const nameSym = expr[1], params = expr[2], body = expr.slice(3);
        if (!(nameSym instanceof ArliSymbol)) throw new SyntaxError(`defn-rec expects a symbol name, got ${nameSym}`);
        if (!Array.isArray(params)) throw new SyntaxError('defn-rec expects a parameter list');
        const fn = new Function(params, body, this.env, nameSym.name);
        this.env.define(nameSym.name, fn);
        this.arityTable.register(nameSym.name, params.length);
        return fn;
      }

      if (head instanceof ArliSymbol && head.name === 'defn-fexpr') {
        if (expr.length < 4) throw new SyntaxError('defn-fexpr expects (defn-fexpr name (params) body...)');
        const nameSym = expr[1], params = expr[2], body = expr.slice(3);
        if (!(nameSym instanceof ArliSymbol)) throw new SyntaxError(`defn-fexpr expects a symbol name, got ${nameSym}`);
        if (!Array.isArray(params)) throw new SyntaxError('defn-fexpr expects a parameter list');
        const fn = new Function(params, body, this.env, nameSym.name, true);
        this.env.define(nameSym.name, fn);
        this.arityTable.register(nameSym.name, params.length);
        return fn;
      }

      // Generic function call
      let fnVal = this._evalExpr(head);
      while (fnVal instanceof ArliSymbol) {
        fnVal = this.env.lookup(fnVal.name);
      }
      if (fnVal instanceof Function && fnVal.isFexpr) return this._applyFunction(fnVal, expr.slice(1));
      const args = expr.slice(1).map(arg => this._evalExpr(arg));
      if (fnVal instanceof Builtin) return fnVal.call(args, this);
      if (fnVal instanceof Function) return this._applyFunction(fnVal, args);
      if (typeof fnVal === 'function') return fnVal(...args);
      throw new TypeError(`Cannot call non-function: ${arliRepr(fnVal)}`);
    }
    throw new TypeError(`Unknown expression type: ${typeof expr}: ${expr}`);
  }

  apply(fn, args) {
    while (fn instanceof ArliSymbol) {
      fn = this.env.lookup(fn.name);
    }
    if (fn instanceof Builtin) return fn.call(args, this);
    if (fn instanceof Function) return this._applyFunction(fn, args);
    if (typeof fn === 'function') return fn(...args);
    throw new TypeError(`Cannot call non-function: ${arliRepr(fn)}`);
  }

  _applyFunction(fn, args) {
    if (args.length !== fn.params.length) throw new TypeError(`Function expected ${fn.params.length} args, got ${args.length}`);
    const callEnv = fn.env.extend(`call(${fn.name || 'anon'})`);
    const oldEnv = this.env;
    this.env = callEnv;
    try {
      for (let i = 0; i < fn.params.length; i++) this.env.define(fn.params[i].name, args[i]);
      const body = Array.isArray(fn.body) ? fn.body : [fn.body];
      let result = nil;
      for (const sub of body) result = this._evalExpr(sub);
      return result;
    } finally { this.env = oldEnv; }
  }

  _matchPattern(pattern, value) {
    if (pattern instanceof ArliSymbol) {
      if (pattern.name === '_') return true;
      if (value instanceof ArliSymbol) return pattern.name === value.name;
      return false;
    }
    if (Array.isArray(pattern) && Array.isArray(value)) {
      if (pattern.length !== value.length) return false;
      for (let i = 0; i < pattern.length; i++) { if (!this._matchPattern(pattern[i], value[i])) return false; }
      return true;
    }
    return arliEqual(pattern, value);
  }

  exec(source) {
    const tokens = tokenize(source);
    const stream = new TokenStream(tokens);
    let result = nil;
    while (!stream.isEOF) {
      const expr = this.parser._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) result = this.eval(expr);
    }
    return result;
  }

  execFile(path) {
    // Node: sync readFile. Browser: fetch + fire exec async, return nil (result not available yet).
    if (nodeFs) {
      return this.exec(nodeFs.readFileSync(path, 'utf-8'));
    }
    // Browser: async fetch, execute when available
    if (typeof fetch !== 'undefined') {
      fetch(path)
        .then(r => { if (!r.ok) throw new Error(`fetch ${path}: ${r.status}`); return r.text(); })
        .then(source => this.exec(source))
        .catch(e => { console.error(`import-module error: ${e.message}`); });
      return nil;
    }
    return nil;
  }
}
